crates/red_knot/src/workspace.rs

```diff
@@ -0,0 +1,328 @@
+// TODO: Fix clippy warnings created by salsa macros
+#![allow(clippy::used_underscore_binding)]
+
+use std::{collections::BTreeMap, sync::Arc};
+
+use rustc_hash::{FxBuildHasher, FxHashSet};
+
+pub use metadata::{PackageMetadata, WorkspaceMetadata};
+use ruff_db::{
+    files::{system_path_to_file, File},
+    system::{walk_directory::WalkState, SystemPath, SystemPathBuf},
+};
+use ruff_python_ast::{name::Name, PySourceType};
+
+use crate::{
+    db::Db,
+    lint::{lint_semantic, lint_syntax, Diagnostics},
+};
+
+mod metadata;
+
+/// The project workspace as a Salsa ingredient.
+///
+/// ## Examples
+///
+/// ```text
+/// app-1/
+///     pyproject.toml
+///     src/
+///         ... python files
+///
+/// app-2/
+///     pyproject.toml
+///     src/
+///         ... python files
+///
+/// shared/
+///     pyproject.toml
+///     src/
+///         ... python files
+///
+/// pyproject.toml
+/// ```
+///
+/// The above project structure has three packages: `app-1`, `app-2`, and `shared`.
+/// Each of the packages can define their own settings in their `pyproject.toml` file, but
```

Comment:
I think yes, "packages" can be nested. 

I intentionally didn't specify how the heuristic works I think it's not important for setting up the relevant concepts. We have to think about how the configuration works and what happens in the absence of any configuration, but I recommend doing this alter and to mirror uv's behavior for best compatibility between all Astral tools. 

My current thinking is:

* 1) Yes packages, can be nested, workspaces can't (or they can, but ruff only picks up the closest workspace and ignores nested workspaces)
* I have no opinion on 2, but I don't see a reason why it can't be a `ruff.toml`.
* 3) We should use Ruff's default configuration and assume the current folder is a workspace containing a single package/project.
* 4) It's unclear to me what the challenge with that is. Yes, this should be supported. The workspace members can be arbitrary paths in the project.

Reply:
How this works in uv: https://github.com/astral-sh/uv/blob/00c055a6bd8a93c980c4d32772b7fd55069de4fe/crates/uv-distribution/src/workspace.rs#L70-L160

Reply:
I expanded the comment on the assumptions that I think are important regardless of the heuristic to discover workspaces. 

* A workspace must always contain at least one package
* Packages can be nested
* Every file belongs to no or exactly one package.